flytekit/clis/sdk_in_container/serialize.py

```diff
@@ -1,50 +1,132 @@
-    tmap = {}
-    for _, _, t in iterate_registerable_entities_in_order(pkgs, include_entities={_sdk_task.SdkTask}):
-        tmap[t.id] = t
+def serialize_all(project, domain, pkgs, version, folder=None):
+    """
+    In order to register, we have to comply with Admin's endpoints. Those endpoints take the following object. These
+    flyteidl.admin.launch_plan_pb2.LaunchPlanSpec
+    flyteidl.admin.workflow_pb2.WorkflowSpec
+    flyteidl.admin.task_pb2.TaskSpec
 
-    for m, k, w in iterate_registerable_entities_in_order(pkgs, include_entities={_workflow.SdkWorkflow}):
-        click.echo('Serializing {}'.format(_utils.fqdn(m.__name__, k, entity_type=w.resource_type)))
-        task_templates = []
-        for n in w.nodes:
-            if n.task_node is not None:
-                task_templates.append(tmap[n.task_node.reference_id])
+    However, if we were to merely call .to_flyte_idl() on all the discovered entities, what we would get are:
+    flyteidl.admin.launch_plan_pb2.LaunchPlanSpec
+    flyteidl.core.workflow_pb2.WorkflowTemplate
+    flyteidl.core.tasks_pb2.TaskTemplate
+
+    For Workflows and Tasks therefore, there is special logic in the serialize function that translates these objects.
+
+    :param Text project:
+    :param Text domain:
+    :param list[Text] pkgs:
+    :param Text version:
+    :param Text folder:
+
+    :return:
+    """
 
-        wc = _WorkflowClosure(workflow=w, tasks=task_templates)
-        wc_pb = wc.to_flyte_idl()
+    # m = module (i.e. python file)
+    # k = value of dir(m), type str
+    # o = object (e.g. SdkWorkflow)
+    loaded_entities = []
+    for m, k, o in iterate_registerable_entities_in_order(pkgs):
+        name = _utils.fqdn(m.__name__, k, entity_type=o.resource_type)
+        _logging.debug("Found module {}\n   K: {} Instantiated in {}".format(m, k, o._instantiated_in))
+        o._id = _identifier.Identifier(
+            o.resource_type,
+            project,
+            domain,
+            name,
+            version
+        )
+        loaded_entities.append(o)
+
+    zero_padded_length = _determine_text_chars(len(loaded_entities))
+    for i, entity in enumerate(loaded_entities):
+        serialized = entity.serialize()
+        fname_index = str(i).zfill(zero_padded_length)
+        fname = '{}_{}.pb'.format(fname_index, entity._id.name)
+        click.echo('  Writing {} to\n    {}'.format(entity._id, fname))
+        _write_proto_to_file(serialized, fname)
+
+        # Not everything serialized will necessarily have an identifier field in it, even though some do (like the
+        # TaskTemplate). To be more rigorous, we write an explicit identifier file that reflects the choices (like
+        # project/domain, etc.) made for this serialize call. We should not allow users to specify a different project
+        # for instance come registration time, to avoid mismatches between potential internal ids like the TaskTemplate
```

Comment:
i think admin populates the id in the task template anyways

Reply:
true.